File: client/src/types/water_meter.ts
export type WaterBillingType = 'meter' | 'flat'

export interface ApiWaterMeter {
  id: string
  room_id: string
  billing_type: WaterBillingType
  reading_date: string
  previous_reading: number | null
  current_reading: number | null
  unit_price: number | null
  flat_amount: number | null
  unit_used: number | null
  total_amount: number
  note: string
  created_at: string
  updated_at: string
  room_number: string
}

export interface CreateWaterMeterPayload {
  room_id: string
  billing_type: WaterBillingType
  reading_date: string
  previous_reading?: number
  current_reading?: number
  unit_price?: number
  flat_amount?: number
  note: string
}

export interface UpdateWaterMeterPayload {
  billing_type?: WaterBillingType
  reading_date?: string
  previous_reading?: number
  current_reading?: number
  unit_price?: number
  flat_amount?: number
  note?: string
}

File: server/internal/domain/water_meter.go
package domain

import (
	"context"
	"time"
)

type WaterBillingType string

const (
	WaterBillingTypeMeter WaterBillingType = "meter"
	WaterBillingTypeFlat  WaterBillingType = "flat"
)

type WaterMeter struct {
	ID              string           `json:"id"`
	RoomID          string           `json:"room_id"`
	BillingType     WaterBillingType `json:"billing_type"`
	ReadingDate     time.Time        `json:"reading_date"`
	PreviousReading *float64         `json:"previous_reading"`
	CurrentReading  *float64         `json:"current_reading"`
	UnitPrice       *float64         `json:"unit_price"`
	FlatAmount      *float64         `json:"flat_amount"`
	Note            string           `json:"note"`
	CreatedAt       time.Time        `json:"created_at"`
	UpdatedAt       time.Time        `json:"updated_at"`
	CreatedBy       string           `json:"created_by"`
	UpdatedBy       string           `json:"updated_by"`
	UpdatedByName   string           `json:"updated_by_name"`
}

type WaterMeterDetail struct {
	WaterMeter
	RoomNumber  string   `json:"room_number"`
	UnitUsed    *float64 `json:"unit_used"`
	TotalAmount float64  `json:"total_amount"`
}

type CreateWaterMeterRequest struct {
	RoomID          string           `json:"room_id"`
	BillingType     WaterBillingType `json:"billing_type"`
	ReadingDate     time.Time        `json:"reading_date"`
	PreviousReading *float64         `json:"previous_reading"`
	CurrentReading  *float64         `json:"current_reading"`
	UnitPrice       *float64         `json:"unit_price"`
	FlatAmount      *float64         `json:"flat_amount"`
	Note            string           `json:"note"`
}

type UpdateWaterMeterRequest struct {
	BillingType     WaterBillingType `json:"billing_type"`
	ReadingDate     time.Time        `json:"reading_date"`
	PreviousReading *float64         `json:"previous_reading"`
	CurrentReading  *float64         `json:"current_reading"`
	UnitPrice       *float64         `json:"unit_price"`
	FlatAmount      *float64         `json:"flat_amount"`
	Note            string           `json:"note"`
}

type WaterMeterRepository interface {
	FindByID(ctx context.Context, id string) (*WaterMeter, error)
	FindDetailByID(ctx context.Context, id string) (*WaterMeterDetail, error)
	List(ctx context.Context, limit, offset int) ([]*WaterMeterDetail, error)
	Count(ctx context.Context) (int, error)
	Create(ctx context.Context, m *WaterMeter) error
	Update(ctx context.Context, m *WaterMeter) error
	Delete(ctx context.Context, id string) error
}
